src/backend/cpu/kernel/unwrap.hpp

```diff
@@ -76,5 +76,67 @@ void unwrap_dim(Param<T> out, CParam<T> in, const dim_t wx, const dim_t wy,
     }
 }
 
+template<typename T, int d>
+void unwrap_dim_dilated(Param<T> out, CParam<T> in, const dim_t wx, const dim_t wy,
+                        const dim_t sx, const dim_t sy, const dim_t px, const dim_t py,
+                        const dim_t dx, const dim_t dy)
+{
+    const T *inPtr = in.get();
+    T *outPtr      = out.get();
+
+    af::dim4 idims    = in.dims();
+    af::dim4 odims    = out.dims();
+    af::dim4 istrides = in.strides();
+    af::dim4 ostrides = out.strides();
+
+    dim_t nx = 1 + (idims[0] + 2 * px - (((wx - 1) * dx) + 1)) / sx;
+
+    for(dim_t w = 0; w < odims[3]; w++) {
+        for(dim_t z = 0; z < odims[2]; z++) {
+
+            dim_t cOut = w * ostrides[3] + z * ostrides[2];
+            dim_t cIn  = w * istrides[3] + z * istrides[2];
+            const T* iptr = inPtr + cIn;
+            T* optr_= outPtr + cOut;
+
+            for(dim_t col = 0; col < odims[d]; col++) {
+                // Offset output ptr
+                T* optr = optr_ + col * ostrides[d];
+
+                // Calculate input window index
+                dim_t winy = (col / nx);
+                dim_t winx = (col % nx);
+
+                dim_t startx = winx * sx;
+                dim_t starty = winy * sy;
+
+                dim_t spx = startx - px;
+                dim_t spy = starty - py;
+
+                // Short cut condition ensuring all values within input dimensions
+                bool cond = (spx >= 0 && spx + (wx * dx) < idims[0] &&
```

Comment:
I doesn't seem like that because `cond` is based on loop index `col` but I believe it can pre-computed as a vector completely out of the batch loop since it seems to depend on `odims[d]`, `nx`, `px` & `py`, `sx` & `sy` which all are loop invariant in computing `cond`.

Reply:
It seems like `cond` calculation has changed since I made my comment. Is that correct ?